### PR TITLE
raster output

### PR DESCRIPTION
**Pull request by michaeldorman**

### CODE REVIEW

File: 08-read-write-plot.qmd [hunk 1]
```diff
@@ -332,7 +335,162 @@ gpd.read_file('output/world_many_layers.gpkg', layer='world2').head(1)
 
 ### Raster data {#sec-data-output-raster}
 
-...
+To write a raster file using `rasterio`, we need to pass a raster file path to `rasterio.open`, in writing (`'w'`) mode. This implies creatign a new empty file (or overwriting an existing one). As opposed to read (`'r'`, the default) mode, the `rasterio.open` function needs quite a lot of information, in addition to the file path and mode:
```
Comment (Nowosad) [inline 08-read-write-plot.qmd:338]: `rasterio` -> **rasterio**

Comment (Nowosad) [reply 08-read-write-plot.qmd:338]: typo "creatign"

File: 08-read-write-plot.qmd [hunk 2]
```diff
@@ -332,7 +335,162 @@ gpd.read_file('output/world_many_layers.gpkg', layer='world2').head(1)
...
+* Metadata describing the raster format and spatial properties
+
+The metadata needs to specify the following properties:
+
+* `driver`---The file format (The recommendation is `'GTiff'` for GeoTIFF)
+* `height`---Number of rows
+* `width`---Number of columns
+* `count`---Number of bands
+* `nodata`---The value which represents “No Data”, if any
+* `dtype`---The raster data type, one of numpy types (e.g., `np.int64`)
```
Comment (Nowosad) [inline 08-read-write-plot.qmd:350]: numpy -> **numpy**

File: 08-read-write-plot.qmd [hunk 3]
```diff
@@ -332,7 +335,162 @@ gpd.read_file('output/world_many_layers.gpkg', layer='world2').head(1)
...
+* `crs`---The CRS, using an EPSG code (e.g., `4326`)
```
Comment (Nowosad) [inline 08-read-write-plot.qmd:351]: Is it possible to use other codes than EPSG?

File: 08-read-write-plot.qmd [hunk 4]
```diff
@@ -332,7 +335,162 @@ gpd.read_file('output/world_many_layers.gpkg', layer='world2').head(1)
...
+* `transform`---The transform matrix
+* `compress`---a compression method to apply, such as `'lzw'`. This is optional and most useful for large rasters. Note that, at the time of writing, this [doesn't work well](https://gis.stackexchange.com/questions/404738/why-does-rasterio-compression-reduces-image-size-with-single-band-but-not-with-m) for writing multiband rasters.
```
Comment (Nowosad) [inline 08-read-write-plot.qmd:353]: a -> A

File: 08-read-write-plot.qmd [hunk 5]
```diff
@@ -332,7 +335,162 @@ gpd.read_file('output/world_many_layers.gpkg', layer='world2').head(1)
...
+
+Once the file connection with the right metadata is ready, we do the actual writing using the `.write` method of the file connection. If there are several bands we may execute the `.write` method several times, as in `.write(a,n)`, where `a` is the array with band values and `n` is the band index (starting from `1`, see below). When done, we close the file connection using the `.close` method. Some functions, such as `rasterio.warp.reproject` used for resampling and reprojecting, directly accept a file connection in `'w'` mode, thus handling the writing (of a resampled or reprojected raster) for us.
```
Comment (Nowosad) [inline 08-read-write-plot.qmd:355]: Maybe it would be good to show a first example directly after this paragraph?

File: 08-read-write-plot.qmd [hunk 6]
```diff
@@ -332,7 +335,162 @@ gpd.read_file('output/world_many_layers.gpkg', layer='world2').head(1)
...
+
+Earlier in the book, we have already demonstrated the four most common scenarios of writing rasters:
+
+* Creating from scratch (@sec-raster-from-scratch)---we created and wrote two rasters from scratch by associating the `elev` and `grain` arrays with an arbitrary spatial extent. The custom arbitrary `transform` created using `rasterio.transform.from_origin`.
+* Aggregating (@sec-raster-agg-disagg)---we wrote an aggregated a raster, by reading a resampled array from an exising raster, then updating the `transform` using `.transform.scale`.
+* Resampling (@sec-raster-resampling)---we resampled a raster into a custom grid, manually creating the `transform` using `rasterio.transform.from_origin`, then resampling and writing the output using `rasterio.warp.reproject`.
+* Reprojecting (@sec-reprojecting-raster-geometries)---We reprojected a raster into another CRS, by automatically calculating an optimal `transform` using `rasterio.warp.calculate_default_transform`, then resampling and writing the output using `rasterio.warp.reproject`.
```
Comment (Nowosad) [inline 08-read-write-plot.qmd:364]: We or we?

File: 08-read-write-plot.qmd [hunk 7]
```diff
@@ -332,7 +335,162 @@ gpd.read_file('output/world_many_layers.gpkg', layer='world2').head(1)
...
+    * Calculate automatically, using `rasterio.warp.calculate_default_transform` (@sec-reprojecting-raster-geometries)
+* The way that the raster is written:
+    * Using the `.write` method, given an existing array (@sec-raster-from-scratch, @sec-raster-agg-disagg)
+    * Using `rasterio.warp.reproject` to calculate and write a resampled or reprojected array (@sec-raster-resampling, @sec-reprojecting-raster-geometries)
+
+To make the picture of raster export complete, there are three important concepts we haven't covered yet: array and raster data types, writing multiband rasters, and handling "No Data" values.
+
+Arrays (i.e., `ndarray` objects defined in package `numpy`) are used to store raster values when reading them from file, suing `.read` (@sec-using-rasterio). All values in an array are of the same type, whereas the `numpy` package supports numerous numeric data types of various precision (and, accordingly, memory footprint). Raster formats, such as GeoTIFF, support exactly the same data types, which means that reading a raster file uses as little RAM as possible. The most relevant types are summarized in @tbl-numpy-data-types.
```
Comment (Nowosad) [inline 08-read-write-plot.qmd:424]: **numpy**

Comment (Nowosad) [reply 08-read-write-plot.qmd:424]: suing? -> using

File: 08-read-write-plot.qmd (hunk 3 above)
Comment (michaeldorman) [reply 08-read-write-plot.qmd:351]: Yes:
https://py.geocompx.org/07-reproj.html#sec-coordinate-reference-systems

File: 08-read-write-plot.qmd (hunk 7 above)
Comment (michaeldorman) [reply 08-read-write-plot.qmd:424]: Fixed, thanks!

**Pull request merged**